From: Stained Glass OS <ke7oxh@gmail.com>
Subject: [PATCH] msvcp60: Implement overflow_error, underflow_error and domain_error.

msvcp60's overflow_error, underflow_error and domain_error: the
constructors from a basic_string, copy constructors, assignments,
destructors, vtables (exported data) and _Doraise (the shared FIXME of
the other msvcp60 exception classes), as the msvcp90 sources build
range_error. Also maps ARM names whose i386 twins are implemented.
---
 dlls/msvcp60/msvcp60.spec | 158 +++++++++++++++++++++++-----------------------
 dlls/msvcp90/exception.c  | 123 ++++++++++++++++++++++++++++++++++--
 2 files changed, 198 insertions(+), 83 deletions(-)

diff --git a/dlls/msvcp60/msvcp60.spec b/dlls/msvcp60/msvcp60.spec
index 4588047..29553eb 100644
--- a/dlls/msvcp60/msvcp60.spec
+++ b/dlls/msvcp60/msvcp60.spec
@@ -503,13 +503,13 @@
 @ cdecl -arch=win32 ??$sqrt@O@std@@YA?AV?$complex@O@0@ABV10@@Z(ptr ptr) complex_double_sqrt
 @ cdecl -arch=win64 ??$sqrt@O@std@@YA?AV?$complex@O@0@AEBV10@@Z(ptr ptr) complex_double_sqrt
 @ cdecl -arch=win64 ??$sqrt@O@std@@YA?AV?$complex@O@1@AEBV21@@Z(ptr ptr) complex_double_sqrt
-@ stub -arch=arm ??0?$_Complex_base@M@std@@QAA@ABM0@Z
+@ cdecl -arch=arm ??0?$_Complex_base@M@std@@QAA@ABM0@Z(ptr ptr ptr) complex_float_ctor
 @ thiscall -arch=i386 ??0?$_Complex_base@M@std@@QAE@ABM0@Z(ptr ptr ptr) complex_float_ctor
 @ cdecl -arch=win64 ??0?$_Complex_base@M@std@@QEAA@AEBM0@Z(ptr ptr ptr) complex_float_ctor
-@ stub -arch=arm ??0?$_Complex_base@N@std@@QAA@ABN0@Z
+@ cdecl -arch=arm ??0?$_Complex_base@N@std@@QAA@ABN0@Z(ptr ptr ptr) complex_double_ctor
 @ thiscall -arch=i386 ??0?$_Complex_base@N@std@@QAE@ABN0@Z(ptr ptr ptr) complex_double_ctor
 @ cdecl -arch=win64 ??0?$_Complex_base@N@std@@QEAA@AEBN0@Z(ptr ptr ptr) complex_double_ctor
-@ stub -arch=arm ??0?$_Complex_base@O@std@@QAA@ABO0@Z
+@ cdecl -arch=arm ??0?$_Complex_base@O@std@@QAA@ABO0@Z(ptr ptr ptr) complex_double_ctor
 @ thiscall -arch=i386 ??0?$_Complex_base@O@std@@QAE@ABO0@Z(ptr ptr ptr) complex_double_ctor
 @ cdecl -arch=win64 ??0?$_Complex_base@O@std@@QEAA@AEBO0@Z(ptr ptr ptr) complex_double_ctor
 @ cdecl -arch=arm ??0?$_Mpunct@D@std@@QAA@ABV_Locinfo@1@I_N@Z(ptr long long) mpunct_char_ctor_refs
@@ -941,13 +941,13 @@
 @ cdecl -arch=arm ??0?$num_put@GV?$ostreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QAA@I@Z(ptr long) num_put_short_ctor_refs
 @ thiscall -arch=i386 ??0?$num_put@GV?$ostreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QAE@I@Z(ptr long) num_put_short_ctor_refs
 @ cdecl -arch=win64 ??0?$num_put@GV?$ostreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QEAA@_K@Z(ptr long) num_put_short_ctor_refs
-@ stub -arch=arm ??0?$numpunct@D@std@@QAA@ABV_Locinfo@1@I@Z
+@ cdecl -arch=arm ??0?$numpunct@D@std@@QAA@ABV_Locinfo@1@I@Z(ptr ptr long) numpunct_char_ctor_locinfo_old
 @ thiscall -arch=i386 ??0?$numpunct@D@std@@QAE@ABV_Locinfo@1@I@Z(ptr ptr long) numpunct_char_ctor_locinfo_old
 @ cdecl -arch=win64 ??0?$numpunct@D@std@@QEAA@AEBV_Locinfo@1@_K@Z(ptr ptr long) numpunct_char_ctor_locinfo_old
 @ cdecl -arch=arm ??0?$numpunct@D@std@@QAA@I@Z(ptr long) numpunct_char_ctor_refs
 @ thiscall -arch=i386 ??0?$numpunct@D@std@@QAE@I@Z(ptr long) numpunct_char_ctor_refs
 @ cdecl -arch=win64 ??0?$numpunct@D@std@@QEAA@_K@Z(ptr long) numpunct_char_ctor_refs
-@ stub -arch=arm ??0?$numpunct@G@std@@QAA@ABV_Locinfo@1@I@Z
+@ cdecl -arch=arm ??0?$numpunct@G@std@@QAA@ABV_Locinfo@1@I@Z(ptr ptr long) numpunct_short_ctor_locinfo_old
 @ thiscall -arch=i386 ??0?$numpunct@G@std@@QAE@ABV_Locinfo@1@I@Z(ptr ptr long) numpunct_short_ctor_locinfo_old
 @ cdecl -arch=win64 ??0?$numpunct@G@std@@QEAA@AEBV_Locinfo@1@_K@Z(ptr ptr long) numpunct_short_ctor_locinfo_old
 @ cdecl -arch=arm ??0?$numpunct@G@std@@QAA@I@Z(ptr long) numpunct_short_ctor_refs
@@ -1040,12 +1040,12 @@
 @ cdecl -arch=arm ??0ctype_base@std@@QAA@I@Z(ptr long) ctype_base_ctor_refs
 @ thiscall -arch=i386 ??0ctype_base@std@@QAE@I@Z(ptr long) ctype_base_ctor_refs
 @ cdecl -arch=win64 ??0ctype_base@std@@QEAA@_K@Z(ptr long) ctype_base_ctor_refs
-@ stub -arch=arm ??0domain_error@std@@QAA@ABV01@@Z
-@ stub -arch=i386 ??0domain_error@std@@QAE@ABV01@@Z
-@ stub -arch=win64 ??0domain_error@std@@QEAA@AEBV01@@Z
-@ stub -arch=arm ??0domain_error@std@@QAA@ABV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@@Z
-@ stub -arch=i386 ??0domain_error@std@@QAE@ABV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@@Z
-@ stub -arch=win64 ??0domain_error@std@@QEAA@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@@Z
+@ cdecl -arch=arm ??0domain_error@std@@QAA@ABV01@@Z(ptr ptr) domain_error_copy_ctor
+@ thiscall -arch=i386 ??0domain_error@std@@QAE@ABV01@@Z(ptr ptr) domain_error_copy_ctor
+@ cdecl -arch=win64 ??0domain_error@std@@QEAA@AEBV01@@Z(ptr ptr) domain_error_copy_ctor
+@ cdecl -arch=arm ??0domain_error@std@@QAA@ABV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@@Z(ptr ptr) MSVCP_domain_error_ctor_bstr
+@ thiscall -arch=i386 ??0domain_error@std@@QAE@ABV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@@Z(ptr ptr) MSVCP_domain_error_ctor_bstr
+@ cdecl -arch=win64 ??0domain_error@std@@QEAA@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@@Z(ptr ptr) MSVCP_domain_error_ctor_bstr
 @ cdecl -arch=arm ??0facet@locale@std@@IAA@I@Z(ptr long) locale_facet_ctor_refs
 @ thiscall -arch=i386 ??0facet@locale@std@@IAE@I@Z(ptr long) locale_facet_ctor_refs
 @ cdecl -arch=win64 ??0facet@locale@std@@IEAA@_K@Z(ptr long) locale_facet_ctor_refs
@@ -1103,12 +1103,12 @@
 @ cdecl -arch=arm ??0out_of_range@std@@QAA@ABV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@@Z(ptr ptr) MSVCP_out_of_range_ctor_bstr
 @ thiscall -arch=i386 ??0out_of_range@std@@QAE@ABV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@@Z(ptr ptr) MSVCP_out_of_range_ctor_bstr
 @ cdecl -arch=win64 ??0out_of_range@std@@QEAA@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@@Z(ptr ptr) MSVCP_out_of_range_ctor_bstr
-@ stub -arch=arm ??0overflow_error@std@@QAA@ABV01@@Z
-@ stub -arch=i386 ??0overflow_error@std@@QAE@ABV01@@Z
-@ stub -arch=win64 ??0overflow_error@std@@QEAA@AEBV01@@Z
-@ stub -arch=arm ??0overflow_error@std@@QAA@ABV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@@Z
-@ stub -arch=i386 ??0overflow_error@std@@QAE@ABV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@@Z
-@ stub -arch=win64 ??0overflow_error@std@@QEAA@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@@Z
+@ cdecl -arch=arm ??0overflow_error@std@@QAA@ABV01@@Z(ptr ptr) overflow_error_copy_ctor
+@ thiscall -arch=i386 ??0overflow_error@std@@QAE@ABV01@@Z(ptr ptr) overflow_error_copy_ctor
+@ cdecl -arch=win64 ??0overflow_error@std@@QEAA@AEBV01@@Z(ptr ptr) overflow_error_copy_ctor
+@ cdecl -arch=arm ??0overflow_error@std@@QAA@ABV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@@Z(ptr ptr) MSVCP_overflow_error_ctor_bstr
+@ thiscall -arch=i386 ??0overflow_error@std@@QAE@ABV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@@Z(ptr ptr) MSVCP_overflow_error_ctor_bstr
+@ cdecl -arch=win64 ??0overflow_error@std@@QEAA@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@@Z(ptr ptr) MSVCP_overflow_error_ctor_bstr
 @ cdecl -arch=arm ??0range_error@std@@QAA@ABV01@@Z(ptr ptr) range_error_copy_ctor
 @ thiscall -arch=i386 ??0range_error@std@@QAE@ABV01@@Z(ptr ptr) range_error_copy_ctor
 @ cdecl -arch=win64 ??0range_error@std@@QEAA@AEBV01@@Z(ptr ptr) range_error_copy_ctor
@@ -1127,12 +1127,12 @@
 @ cdecl -arch=arm ??0time_base@std@@QAA@I@Z(ptr long) time_base_ctor_refs
 @ thiscall -arch=i386 ??0time_base@std@@QAE@I@Z(ptr long) time_base_ctor_refs
 @ cdecl -arch=win64 ??0time_base@std@@QEAA@_K@Z(ptr long) time_base_ctor_refs
-@ stub -arch=arm ??0underflow_error@std@@QAA@ABV01@@Z
-@ stub -arch=i386 ??0underflow_error@std@@QAE@ABV01@@Z
-@ stub -arch=win64 ??0underflow_error@std@@QEAA@AEBV01@@Z
-@ stub -arch=arm ??0underflow_error@std@@QAA@ABV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@@Z
-@ stub -arch=i386 ??0underflow_error@std@@QAE@ABV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@@Z
-@ stub -arch=win64 ??0underflow_error@std@@QEAA@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@@Z
+@ cdecl -arch=arm ??0underflow_error@std@@QAA@ABV01@@Z(ptr ptr) underflow_error_copy_ctor
+@ thiscall -arch=i386 ??0underflow_error@std@@QAE@ABV01@@Z(ptr ptr) underflow_error_copy_ctor
+@ cdecl -arch=win64 ??0underflow_error@std@@QEAA@AEBV01@@Z(ptr ptr) underflow_error_copy_ctor
+@ cdecl -arch=arm ??0underflow_error@std@@QAA@ABV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@@Z(ptr ptr) MSVCP_underflow_error_ctor_bstr
+@ thiscall -arch=i386 ??0underflow_error@std@@QAE@ABV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@@Z(ptr ptr) MSVCP_underflow_error_ctor_bstr
+@ cdecl -arch=win64 ??0underflow_error@std@@QEAA@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@@Z(ptr ptr) MSVCP_underflow_error_ctor_bstr
 @ cdecl -arch=arm ??1?$_Mpunct@D@std@@UAA@XZ(ptr) mpunct_char_dtor
 @ thiscall -arch=i386 ??1?$_Mpunct@D@std@@UAE@XZ(ptr) mpunct_char_dtor
 @ cdecl -arch=win64 ??1?$_Mpunct@D@std@@UEAA@XZ(ptr) mpunct_char_dtor
@@ -1289,10 +1289,10 @@
 @ cdecl -arch=arm ??1?$numpunct@G@std@@UAA@XZ(ptr) numpunct_wchar_dtor
 @ thiscall -arch=i386 ??1?$numpunct@G@std@@UAE@XZ(ptr) numpunct_wchar_dtor
 @ cdecl -arch=win64 ??1?$numpunct@G@std@@UEAA@XZ(ptr) numpunct_wchar_dtor
-@ stub -arch=arm ??1?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@UAA@XZ
+@ cdecl -arch=arm ??1?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@UAA@XZ(ptr) time_get_char_dtor
 @ thiscall -arch=i386 ??1?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@UAE@XZ(ptr) time_get_char_dtor
 @ cdecl -arch=win64 ??1?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@UEAA@XZ(ptr) time_get_char_dtor
-@ stub -arch=arm ??1?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@UAA@XZ
+@ cdecl -arch=arm ??1?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@UAA@XZ(ptr) time_get_wchar_dtor
 @ thiscall -arch=i386 ??1?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@UAE@XZ(ptr) time_get_wchar_dtor
 @ cdecl -arch=win64 ??1?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@UEAA@XZ(ptr) time_get_wchar_dtor
 @ cdecl -arch=arm ??1?$time_put@DV?$ostreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@UAA@XZ(ptr) time_put_char_dtor
@@ -1337,9 +1337,9 @@
 @ cdecl -arch=arm ??1ctype_base@std@@UAA@XZ(ptr) ctype_base_dtor
 @ thiscall -arch=i386 ??1ctype_base@std@@UAE@XZ(ptr) ctype_base_dtor
 @ cdecl -arch=win64 ??1ctype_base@std@@UEAA@XZ(ptr) ctype_base_dtor
-@ stub -arch=arm ??1domain_error@std@@UAA@XZ
-@ stub -arch=i386 ??1domain_error@std@@UAE@XZ
-@ stub -arch=win64 ??1domain_error@std@@UEAA@XZ
+@ cdecl -arch=arm ??1domain_error@std@@UAA@XZ(ptr) MSVCP_logic_error_dtor
+@ thiscall -arch=i386 ??1domain_error@std@@UAE@XZ(ptr) MSVCP_logic_error_dtor
+@ cdecl -arch=win64 ??1domain_error@std@@UEAA@XZ(ptr) MSVCP_logic_error_dtor
 @ cdecl -arch=arm ??1facet@locale@std@@UAA@XZ(ptr) locale_facet_dtor
 @ thiscall -arch=i386 ??1facet@locale@std@@UAE@XZ(ptr) locale_facet_dtor
 @ cdecl -arch=win64 ??1facet@locale@std@@UEAA@XZ(ptr) locale_facet_dtor
@@ -1370,9 +1370,9 @@
 @ cdecl -arch=arm ??1out_of_range@std@@UAA@XZ(ptr) MSVCP_logic_error_dtor
 @ thiscall -arch=i386 ??1out_of_range@std@@UAE@XZ(ptr) MSVCP_logic_error_dtor
 @ cdecl -arch=win64 ??1out_of_range@std@@UEAA@XZ(ptr) MSVCP_logic_error_dtor
-@ stub -arch=arm ??1overflow_error@std@@UAA@XZ
-@ stub -arch=i386 ??1overflow_error@std@@UAE@XZ
-@ stub -arch=win64 ??1overflow_error@std@@UEAA@XZ
+@ cdecl -arch=arm ??1overflow_error@std@@UAA@XZ(ptr) MSVCP_runtime_error_dtor
+@ thiscall -arch=i386 ??1overflow_error@std@@UAE@XZ(ptr) MSVCP_runtime_error_dtor
+@ cdecl -arch=win64 ??1overflow_error@std@@UEAA@XZ(ptr) MSVCP_runtime_error_dtor
 @ cdecl -arch=arm ??1range_error@std@@UAA@XZ(ptr) MSVCP_runtime_error_dtor
 @ thiscall -arch=i386 ??1range_error@std@@UAE@XZ(ptr) MSVCP_runtime_error_dtor
 @ cdecl -arch=win64 ??1range_error@std@@UEAA@XZ(ptr) MSVCP_runtime_error_dtor
@@ -1388,25 +1388,25 @@
 @ cdecl -arch=arm ??1time_base@std@@UAA@XZ(ptr) time_base_dtor
 @ thiscall -arch=i386 ??1time_base@std@@UAE@XZ(ptr) time_base_dtor
 @ cdecl -arch=win64 ??1time_base@std@@UEAA@XZ(ptr) time_base_dtor
-@ stub -arch=arm ??1underflow_error@std@@UAA@XZ
-@ stub -arch=i386 ??1underflow_error@std@@UAE@XZ
-@ stub -arch=win64 ??1underflow_error@std@@UEAA@XZ
-@ stub -arch=arm ??4?$_Complex_base@M@std@@QAAAAV01@ABV01@@Z
+@ cdecl -arch=arm ??1underflow_error@std@@UAA@XZ(ptr) MSVCP_runtime_error_dtor
+@ thiscall -arch=i386 ??1underflow_error@std@@UAE@XZ(ptr) MSVCP_runtime_error_dtor
+@ cdecl -arch=win64 ??1underflow_error@std@@UEAA@XZ(ptr) MSVCP_runtime_error_dtor
+@ cdecl -arch=arm ??4?$_Complex_base@M@std@@QAAAAV01@ABV01@@Z(ptr ptr) complex_float_assign
 @ thiscall -arch=i386 ??4?$_Complex_base@M@std@@QAEAAV01@ABV01@@Z(ptr ptr) complex_float_assign
 @ cdecl -arch=win64 ??4?$_Complex_base@M@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) complex_float_assign
-@ stub -arch=arm ??4?$_Complex_base@N@std@@QAAAAV01@ABV01@@Z
+@ cdecl -arch=arm ??4?$_Complex_base@N@std@@QAAAAV01@ABV01@@Z(ptr ptr) complex_double_assign
 @ thiscall -arch=i386 ??4?$_Complex_base@N@std@@QAEAAV01@ABV01@@Z(ptr ptr) complex_double_assign
 @ cdecl -arch=win64 ??4?$_Complex_base@N@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) complex_double_assign
-@ stub -arch=arm ??4?$_Complex_base@O@std@@QAAAAV01@ABV01@@Z
+@ cdecl -arch=arm ??4?$_Complex_base@O@std@@QAAAAV01@ABV01@@Z(ptr ptr) complex_double_assign
 @ thiscall -arch=i386 ??4?$_Complex_base@O@std@@QAEAAV01@ABV01@@Z(ptr ptr) complex_double_assign
 @ cdecl -arch=win64 ??4?$_Complex_base@O@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) complex_double_assign
-@ stub -arch=arm ??4?$_Ctr@M@std@@QAAAAV01@ABV01@@Z
+@ cdecl -arch=arm ??4?$_Ctr@M@std@@QAAAAV01@ABV01@@Z(ptr ptr) std_Ctraits_op_assign
 @ thiscall -arch=i386 ??4?$_Ctr@M@std@@QAEAAV01@ABV01@@Z(ptr ptr) std_Ctraits_op_assign
 @ cdecl -arch=win64 ??4?$_Ctr@M@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) std_Ctraits_op_assign
-@ stub -arch=arm ??4?$_Ctr@N@std@@QAAAAV01@ABV01@@Z
+@ cdecl -arch=arm ??4?$_Ctr@N@std@@QAAAAV01@ABV01@@Z(ptr ptr) std_Ctraits_op_assign
 @ thiscall -arch=i386 ??4?$_Ctr@N@std@@QAEAAV01@ABV01@@Z(ptr ptr) std_Ctraits_op_assign
 @ cdecl -arch=win64 ??4?$_Ctr@N@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) std_Ctraits_op_assign
-@ stub -arch=arm ??4?$_Ctr@O@std@@QAAAAV01@ABV01@@Z
+@ cdecl -arch=arm ??4?$_Ctr@O@std@@QAAAAV01@ABV01@@Z(ptr ptr) std_Ctraits_op_assign
 @ thiscall -arch=i386 ??4?$_Ctr@O@std@@QAEAAV01@ABV01@@Z(ptr ptr) std_Ctraits_op_assign
 @ cdecl -arch=win64 ??4?$_Ctr@O@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) std_Ctraits_op_assign
 @ cdecl -arch=arm ??4?$allocator@X@std@@QAAAAV01@ABV01@@Z(ptr ptr) MSVCP_allocator_void_assign
@@ -1610,9 +1610,9 @@
 @ stub -arch=arm ??4bad_typeid@std@@QAAAAV01@ABV01@@Z
 @ stub -arch=i386 ??4bad_typeid@std@@QAEAAV01@ABV01@@Z
 @ stub -arch=win64 ??4bad_typeid@std@@QEAAAEAV01@AEBV01@@Z
-@ stub -arch=arm ??4domain_error@std@@QAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4domain_error@std@@QAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4domain_error@std@@QEAAAEAV01@AEBV01@@Z
+@ cdecl -arch=arm ??4domain_error@std@@QAAAAV01@ABV01@@Z(ptr ptr) MSVCP_domain_error_assign
+@ thiscall -arch=i386 ??4domain_error@std@@QAEAAV01@ABV01@@Z(ptr ptr) MSVCP_domain_error_assign
+@ cdecl -arch=win64 ??4domain_error@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) MSVCP_domain_error_assign
 @ stub -arch=arm ??4id@locale@std@@QAAAAV012@ABV012@@Z
 @ stub -arch=i386 ??4id@locale@std@@QAEAAV012@ABV012@@Z
 @ stub -arch=win64 ??4id@locale@std@@QEAAAEAV012@AEBV012@@Z
@@ -1631,18 +1631,18 @@
 @ cdecl -arch=arm ??4out_of_range@std@@QAAAAV01@ABV01@@Z(ptr ptr) MSVCP_out_of_range_assign
 @ thiscall -arch=i386 ??4out_of_range@std@@QAEAAV01@ABV01@@Z(ptr ptr) MSVCP_out_of_range_assign
 @ cdecl -arch=win64 ??4out_of_range@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) MSVCP_out_of_range_assign
-@ stub -arch=arm ??4overflow_error@std@@QAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4overflow_error@std@@QAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4overflow_error@std@@QEAAAEAV01@AEBV01@@Z
+@ cdecl -arch=arm ??4overflow_error@std@@QAAAAV01@ABV01@@Z(ptr ptr) MSVCP_overflow_error_assign
+@ thiscall -arch=i386 ??4overflow_error@std@@QAEAAV01@ABV01@@Z(ptr ptr) MSVCP_overflow_error_assign
+@ cdecl -arch=win64 ??4overflow_error@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) MSVCP_overflow_error_assign
 @ cdecl -arch=arm ??4range_error@std@@QAAAAV01@ABV01@@Z(ptr ptr) MSVCP_range_error_assign
 @ thiscall -arch=i386 ??4range_error@std@@QAEAAV01@ABV01@@Z(ptr ptr) MSVCP_range_error_assign
 @ cdecl -arch=win64 ??4range_error@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) MSVCP_range_error_assign
 @ cdecl -arch=arm ??4runtime_error@std@@QAAAAV01@ABV01@@Z(ptr ptr) MSVCP_runtime_error_assign
 @ thiscall -arch=i386 ??4runtime_error@std@@QAEAAV01@ABV01@@Z(ptr ptr) MSVCP_runtime_error_assign
 @ cdecl -arch=win64 ??4runtime_error@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) MSVCP_runtime_error_assign
-@ stub -arch=arm ??4underflow_error@std@@QAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4underflow_error@std@@QAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4underflow_error@std@@QEAAAEAV01@AEBV01@@Z
+@ cdecl -arch=arm ??4underflow_error@std@@QAAAAV01@ABV01@@Z(ptr ptr) MSVCP_underflow_error_assign
+@ thiscall -arch=i386 ??4underflow_error@std@@QAEAAV01@ABV01@@Z(ptr ptr) MSVCP_underflow_error_assign
+@ cdecl -arch=win64 ??4underflow_error@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) MSVCP_underflow_error_assign
 @ cdecl -arch=arm ??5?$basic_istream@DU?$char_traits@D@std@@@std@@QAAAAV01@AAF@Z(ptr ptr) basic_istream_char_read_short
 @ thiscall -arch=i386 ??5?$basic_istream@DU?$char_traits@D@std@@@std@@QAEAAV01@AAF@Z(ptr ptr) basic_istream_char_read_short
 @ cdecl -arch=win64 ??5?$basic_istream@DU?$char_traits@D@std@@@std@@QEAAAEAV01@AEAF@Z(ptr ptr) basic_istream_char_read_short
@@ -2296,7 +2296,7 @@
 # extern ??_7bad_typeid@std@@6B@
 @ extern ??_7codecvt_base@std@@6B@ codecvt_base_vtable
 @ extern ??_7ctype_base@std@@6B@ ctype_base_vtable
-# extern ??_7domain_error@std@@6B@
+@ extern ??_7domain_error@std@@6B@ domain_error_vtable
 @ extern ??_7facet@locale@std@@6B@ locale_facet_vtable
 @ extern ??_7ios_base@std@@6B@ ios_base_vtable
 @ extern ??_7length_error@std@@6B@ length_error_vtable
@@ -2304,11 +2304,11 @@
 @ extern ??_7messages_base@std@@6B@ messages_base_vtable
 @ extern ??_7money_base@std@@6B@ money_base_vtable
 @ extern ??_7out_of_range@std@@6B@ out_of_range_vtable
-# extern ??_7overflow_error@std@@6B@
+@ extern ??_7overflow_error@std@@6B@ overflow_error_vtable
 @ extern ??_7range_error@std@@6B@ range_error_vtable
 @ extern ??_7runtime_error@std@@6B@ runtime_error_vtable
 # extern ??_7time_base@std@@6B@
-# extern ??_7underflow_error@std@@6B@
+@ extern ??_7underflow_error@std@@6B@ underflow_error_vtable
 @ extern ??_8?$basic_fstream@DU?$char_traits@D@std@@@std@@7B?$basic_istream@DU?$char_traits@D@std@@@1@@ basic_fstream_char_vbtable1
 @ extern ??_8?$basic_fstream@DU?$char_traits@D@std@@@std@@7B?$basic_ostream@DU?$char_traits@D@std@@@1@@ basic_fstream_char_vbtable2
 @ extern ??_8?$basic_fstream@GU?$char_traits@G@std@@@std@@7B?$basic_istream@GU?$char_traits@G@std@@@1@@ basic_fstream_short_vbtable1
@@ -2592,9 +2592,9 @@
 @ stub -arch=arm ?_Doraise@bad_typeid@std@@MBAXXZ
 @ stub -arch=i386 ?_Doraise@bad_typeid@std@@MBEXXZ
 @ stub -arch=win64 ?_Doraise@bad_typeid@std@@MEBAXXZ
-@ stub -arch=arm ?_Doraise@domain_error@std@@MBAXXZ
-@ stub -arch=i386 ?_Doraise@domain_error@std@@MBEXXZ
-@ stub -arch=win64 ?_Doraise@domain_error@std@@MEBAXXZ
+@ cdecl -arch=arm ?_Doraise@domain_error@std@@MBAXXZ(ptr) MSVCP_exception__Doraise
+@ thiscall -arch=i386 ?_Doraise@domain_error@std@@MBEXXZ(ptr) MSVCP_exception__Doraise
+@ cdecl -arch=win64 ?_Doraise@domain_error@std@@MEBAXXZ(ptr) MSVCP_exception__Doraise
 @ cdecl -arch=arm ?_Doraise@length_error@std@@MBAXXZ(ptr) MSVCP_exception__Doraise
 @ thiscall -arch=i386 ?_Doraise@length_error@std@@MBEXXZ(ptr) MSVCP_exception__Doraise
 @ cdecl -arch=win64 ?_Doraise@length_error@std@@MEBAXXZ(ptr) MSVCP_exception__Doraise
@@ -2604,18 +2604,18 @@
 @ cdecl -arch=arm ?_Doraise@out_of_range@std@@MBAXXZ(ptr) MSVCP_exception__Doraise
 @ thiscall -arch=i386 ?_Doraise@out_of_range@std@@MBEXXZ(ptr) MSVCP_exception__Doraise
 @ cdecl -arch=win64 ?_Doraise@out_of_range@std@@MEBAXXZ(ptr) MSVCP_exception__Doraise
-@ stub -arch=arm ?_Doraise@overflow_error@std@@MBAXXZ
-@ stub -arch=i386 ?_Doraise@overflow_error@std@@MBEXXZ
-@ stub -arch=win64 ?_Doraise@overflow_error@std@@MEBAXXZ
+@ cdecl -arch=arm ?_Doraise@overflow_error@std@@MBAXXZ(ptr) MSVCP_exception__Doraise
+@ thiscall -arch=i386 ?_Doraise@overflow_error@std@@MBEXXZ(ptr) MSVCP_exception__Doraise
+@ cdecl -arch=win64 ?_Doraise@overflow_error@std@@MEBAXXZ(ptr) MSVCP_exception__Doraise
 @ cdecl -arch=arm ?_Doraise@range_error@std@@MBAXXZ(ptr) MSVCP_exception__Doraise
 @ thiscall -arch=i386 ?_Doraise@range_error@std@@MBEXXZ(ptr) MSVCP_exception__Doraise
 @ cdecl -arch=win64 ?_Doraise@range_error@std@@MEBAXXZ(ptr) MSVCP_exception__Doraise
 @ cdecl -arch=arm ?_Doraise@runtime_error@std@@MBAXXZ(ptr) MSVCP_exception__Doraise
 @ thiscall -arch=i386 ?_Doraise@runtime_error@std@@MBEXXZ(ptr) MSVCP_exception__Doraise
 @ cdecl -arch=win64 ?_Doraise@runtime_error@std@@MEBAXXZ(ptr) MSVCP_exception__Doraise
-@ stub -arch=arm ?_Doraise@underflow_error@std@@MBAXXZ
-@ stub -arch=i386 ?_Doraise@underflow_error@std@@MBEXXZ
-@ stub -arch=win64 ?_Doraise@underflow_error@std@@MEBAXXZ
+@ cdecl -arch=arm ?_Doraise@underflow_error@std@@MBAXXZ(ptr) MSVCP_exception__Doraise
+@ thiscall -arch=i386 ?_Doraise@underflow_error@std@@MBEXXZ(ptr) MSVCP_exception__Doraise
+@ cdecl -arch=win64 ?_Doraise@underflow_error@std@@MEBAXXZ(ptr) MSVCP_exception__Doraise
 @ cdecl -arch=arm ?_Eos@?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@AAAXI@Z(ptr long) basic_string_char__Eos
 @ thiscall -arch=i386 ?_Eos@?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@AAEXI@Z(ptr long) basic_string_char__Eos
 @ cdecl -arch=win64 ?_Eos@?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@AEAAX_K@Z(ptr long) basic_string_char__Eos
@@ -2882,10 +2882,10 @@
 @ cdecl -arch=arm ?_Init@?$num_put@GV?$ostreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@IAAXABV_Locinfo@2@@Z(ptr ptr) num_put_wchar__Init
 @ thiscall -arch=i386 ?_Init@?$num_put@GV?$ostreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@IAEXABV_Locinfo@2@@Z(ptr ptr) num_put_wchar__Init
 @ cdecl -arch=win64 ?_Init@?$num_put@GV?$ostreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@IEAAXAEBV_Locinfo@2@@Z(ptr ptr) num_put_wchar__Init
-@ stub -arch=arm ?_Init@?$numpunct@D@std@@IAAXABV_Locinfo@2@@Z
+@ cdecl -arch=arm ?_Init@?$numpunct@D@std@@IAAXABV_Locinfo@2@@Z(ptr ptr) numpunct_char__Init_old
 @ thiscall -arch=i386 ?_Init@?$numpunct@D@std@@IAEXABV_Locinfo@2@@Z(ptr ptr) numpunct_char__Init_old
 @ cdecl -arch=win64 ?_Init@?$numpunct@D@std@@IEAAXAEBV_Locinfo@2@@Z(ptr ptr) numpunct_char__Init_old
-@ stub -arch=arm ?_Init@?$numpunct@G@std@@IAAXABV_Locinfo@2@@Z
+@ cdecl -arch=arm ?_Init@?$numpunct@G@std@@IAAXABV_Locinfo@2@@Z(ptr ptr) numpunct_wchar__Init_old
 @ thiscall -arch=i386 ?_Init@?$numpunct@G@std@@IAEXABV_Locinfo@2@@Z(ptr ptr) numpunct_wchar__Init_old
 @ cdecl -arch=win64 ?_Init@?$numpunct@G@std@@IEAAXAEBV_Locinfo@2@@Z(ptr ptr) numpunct_wchar__Init_old
 @ cdecl -arch=arm ?_Init@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@IAAXABV_Locinfo@2@@Z(ptr ptr) time_get_char__Init
@@ -3342,10 +3342,10 @@
 @ cdecl -arch=arm ?data@?$basic_string@GU?$char_traits@G@std@@V?$allocator@G@2@@std@@QBAPBGXZ(ptr) MSVCP_basic_string_wchar_c_str
 @ thiscall -arch=i386 ?data@?$basic_string@GU?$char_traits@G@std@@V?$allocator@G@2@@std@@QBEPBGXZ(ptr) MSVCP_basic_string_wchar_c_str
 @ cdecl -arch=win64 ?data@?$basic_string@GU?$char_traits@G@std@@V?$allocator@G@2@@std@@QEBAPEBGXZ(ptr) MSVCP_basic_string_wchar_c_str
-@ stub -arch=arm ?date_order@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@QBAHXZ
+@ cdecl -arch=arm ?date_order@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@QBAHXZ(ptr) time_get_char_date_order
 @ thiscall -arch=i386 ?date_order@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@QBEHXZ(ptr) time_get_char_date_order
 @ cdecl -arch=win64 ?date_order@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@QEBAHXZ(ptr) time_get_char_date_order
-@ stub -arch=arm ?date_order@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QBAHXZ
+@ cdecl -arch=arm ?date_order@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QBAHXZ(ptr) time_get_wchar_date_order
 @ thiscall -arch=i386 ?date_order@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QBEHXZ(ptr) time_get_wchar_date_order
 @ cdecl -arch=win64 ?date_order@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@QEBAHXZ(ptr) time_get_wchar_date_order
 @ cdecl -arch=arm ?decimal_point@?$_Mpunct@D@std@@QBADXZ(ptr) mpunct_char_decimal_point
@@ -3397,10 +3397,10 @@
 @ cdecl -arch=arm ?do_curr_symbol@?$_Mpunct@G@std@@MBA?AV?$basic_string@GU?$char_traits@G@std@@V?$allocator@G@2@@2@XZ(ptr ptr) mpunct_short_do_curr_symbol
 @ thiscall -arch=i386 ?do_curr_symbol@?$_Mpunct@G@std@@MBE?AV?$basic_string@GU?$char_traits@G@std@@V?$allocator@G@2@@2@XZ(ptr ptr) mpunct_short_do_curr_symbol
 @ cdecl -arch=win64 ?do_curr_symbol@?$_Mpunct@G@std@@MEBA?AV?$basic_string@GU?$char_traits@G@std@@V?$allocator@G@2@@2@XZ(ptr ptr) mpunct_short_do_curr_symbol
-@ stub -arch=arm ?do_date_order@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@MBAHXZ
+@ cdecl -arch=arm ?do_date_order@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@MBAHXZ(ptr) time_get_char_do_date_order
 @ thiscall -arch=i386 ?do_date_order@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@MBEHXZ(ptr) time_get_char_do_date_order
 @ cdecl -arch=win64 ?do_date_order@?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@MEBAHXZ(ptr) time_get_char_do_date_order
-@ stub -arch=arm ?do_date_order@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@MBAHXZ
+@ cdecl -arch=arm ?do_date_order@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@MBAHXZ(ptr) time_get_wchar_do_date_order
 @ thiscall -arch=i386 ?do_date_order@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@MBEHXZ(ptr) time_get_wchar_do_date_order
 @ cdecl -arch=win64 ?do_date_order@?$time_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@MEBAHXZ(ptr) time_get_wchar_do_date_order
 @ cdecl -arch=arm ?do_decimal_point@?$_Mpunct@D@std@@MBADXZ(ptr) mpunct_char_do_decimal_point
@@ -4271,22 +4271,22 @@
 @ cdecl -arch=arm ?ignore@?$basic_istream@GU?$char_traits@G@std@@@std@@QAAAAV12@HG@Z(ptr long long) basic_istream_short_ignore
 @ thiscall -arch=i386 ?ignore@?$basic_istream@GU?$char_traits@G@std@@@std@@QAEAAV12@HG@Z(ptr long long) basic_istream_short_ignore
 @ cdecl -arch=win64 ?ignore@?$basic_istream@GU?$char_traits@G@std@@@std@@QEAAAEAV12@_JG@Z(ptr long long) basic_istream_short_ignore
-@ stub -arch=arm ?imag@?$_Complex_base@M@std@@QAAMABM@Z
+@ cdecl -arch=arm ?imag@?$_Complex_base@M@std@@QAAMABM@Z(ptr ptr) complex_float_imag_set
 @ thiscall -arch=i386 ?imag@?$_Complex_base@M@std@@QAEMABM@Z(ptr ptr) complex_float_imag_set
 @ cdecl -arch=win64 ?imag@?$_Complex_base@M@std@@QEAAMAEBM@Z(ptr ptr) complex_float_imag_set
-@ stub -arch=arm ?imag@?$_Complex_base@M@std@@QBAMXZ
+@ cdecl -arch=arm ?imag@?$_Complex_base@M@std@@QBAMXZ(ptr) complex_float_imag_get
 @ thiscall -arch=i386 ?imag@?$_Complex_base@M@std@@QBEMXZ(ptr) complex_float_imag_get
 @ cdecl -arch=win64 ?imag@?$_Complex_base@M@std@@QEBAMXZ(ptr) complex_float_imag_get
-@ stub -arch=arm ?imag@?$_Complex_base@N@std@@QAANABN@Z
+@ cdecl -arch=arm ?imag@?$_Complex_base@N@std@@QAANABN@Z(ptr ptr) complex_double_imag_set
 @ thiscall -arch=i386 ?imag@?$_Complex_base@N@std@@QAENABN@Z(ptr ptr) complex_double_imag_set
 @ cdecl -arch=win64 ?imag@?$_Complex_base@N@std@@QEAANAEBN@Z(ptr ptr) complex_double_imag_set
-@ stub -arch=arm ?imag@?$_Complex_base@N@std@@QBANXZ
+@ cdecl -arch=arm ?imag@?$_Complex_base@N@std@@QBANXZ(ptr) complex_double_imag_get
 @ thiscall -arch=i386 ?imag@?$_Complex_base@N@std@@QBENXZ(ptr) complex_double_imag_get
 @ cdecl -arch=win64 ?imag@?$_Complex_base@N@std@@QEBANXZ(ptr) complex_double_imag_get
-@ stub -arch=arm ?imag@?$_Complex_base@O@std@@QAAOABO@Z
+@ cdecl -arch=arm ?imag@?$_Complex_base@O@std@@QAAOABO@Z(ptr ptr) complex_double_imag_set
 @ thiscall -arch=i386 ?imag@?$_Complex_base@O@std@@QAEOABO@Z(ptr ptr) complex_double_imag_set
 @ cdecl -arch=win64 ?imag@?$_Complex_base@O@std@@QEAAOAEBO@Z(ptr ptr) complex_double_imag_set
-@ stub -arch=arm ?imag@?$_Complex_base@O@std@@QBAOXZ
+@ cdecl -arch=arm ?imag@?$_Complex_base@O@std@@QBAOXZ(ptr) complex_double_imag_get
 @ thiscall -arch=i386 ?imag@?$_Complex_base@O@std@@QBEOXZ(ptr) complex_double_imag_get
 @ cdecl -arch=win64 ?imag@?$_Complex_base@O@std@@QEBAOXZ(ptr) complex_double_imag_get
 @ cdecl -arch=win32 ?imag@std@@YAMABV?$complex@M@1@@Z(ptr) complex_float_imag
@@ -4973,22 +4973,22 @@
 @ cdecl -arch=arm ?readsome@?$basic_istream@GU?$char_traits@G@std@@@std@@QAAHPAGH@Z(ptr ptr long) basic_istream_short_readsome
 @ thiscall -arch=i386 ?readsome@?$basic_istream@GU?$char_traits@G@std@@@std@@QAEHPAGH@Z(ptr ptr long) basic_istream_short_readsome
 @ cdecl -arch=win64 ?readsome@?$basic_istream@GU?$char_traits@G@std@@@std@@QEAA_JPEAG_J@Z(ptr ptr long) basic_istream_short_readsome
-@ stub -arch=arm ?real@?$_Complex_base@M@std@@QAAMABM@Z
+@ cdecl -arch=arm ?real@?$_Complex_base@M@std@@QAAMABM@Z(ptr ptr) complex_float_real_set
 @ thiscall -arch=i386 ?real@?$_Complex_base@M@std@@QAEMABM@Z(ptr ptr) complex_float_real_set
 @ cdecl -arch=win64 ?real@?$_Complex_base@M@std@@QEAAMAEBM@Z(ptr ptr) complex_float_real_set
-@ stub -arch=arm ?real@?$_Complex_base@M@std@@QBAMXZ
+@ cdecl -arch=arm ?real@?$_Complex_base@M@std@@QBAMXZ(ptr) complex_float_real_get
 @ thiscall -arch=i386 ?real@?$_Complex_base@M@std@@QBEMXZ(ptr) complex_float_real_get
 @ cdecl -arch=win64 ?real@?$_Complex_base@M@std@@QEBAMXZ(ptr) complex_float_real_get
-@ stub -arch=arm ?real@?$_Complex_base@N@std@@QAANABN@Z
+@ cdecl -arch=arm ?real@?$_Complex_base@N@std@@QAANABN@Z(ptr ptr) complex_double_real_set
 @ thiscall -arch=i386 ?real@?$_Complex_base@N@std@@QAENABN@Z(ptr ptr) complex_double_real_set
 @ cdecl -arch=win64 ?real@?$_Complex_base@N@std@@QEAANAEBN@Z(ptr ptr) complex_double_real_set
-@ stub -arch=arm ?real@?$_Complex_base@N@std@@QBANXZ
+@ cdecl -arch=arm ?real@?$_Complex_base@N@std@@QBANXZ(ptr) complex_double_real_get
 @ thiscall -arch=i386 ?real@?$_Complex_base@N@std@@QBENXZ(ptr) complex_double_real_get
 @ cdecl -arch=win64 ?real@?$_Complex_base@N@std@@QEBANXZ(ptr) complex_double_real_get
-@ stub -arch=arm ?real@?$_Complex_base@O@std@@QAAOABO@Z
+@ cdecl -arch=arm ?real@?$_Complex_base@O@std@@QAAOABO@Z(ptr ptr) complex_double_real_set
 @ thiscall -arch=i386 ?real@?$_Complex_base@O@std@@QAEOABO@Z(ptr ptr) complex_double_real_set
 @ cdecl -arch=win64 ?real@?$_Complex_base@O@std@@QEAAOAEBO@Z(ptr ptr) complex_double_real_set
-@ stub -arch=arm ?real@?$_Complex_base@O@std@@QBAOXZ
+@ cdecl -arch=arm ?real@?$_Complex_base@O@std@@QBAOXZ(ptr) complex_double_real_get
 @ thiscall -arch=i386 ?real@?$_Complex_base@O@std@@QBEOXZ(ptr) complex_double_real_get
 @ cdecl -arch=win64 ?real@?$_Complex_base@O@std@@QEBAOXZ(ptr) complex_double_real_get
 @ cdecl -arch=win32 ?real@std@@YAMABV?$complex@M@1@@Z(ptr) complex_float_real
diff --git a/dlls/msvcp90/exception.c b/dlls/msvcp90/exception.c
index cce20a7..8dd680e 100644
--- a/dlls/msvcp90/exception.c
+++ b/dlls/msvcp90/exception.c
@@ -886,7 +886,7 @@ range_error* __thiscall MSVCP_range_error_assign(range_error *this, const range_
 DEFINE_RTTI_DATA2(range_error, 0, &runtime_error_rtti_base_descriptor, &exception_rtti_base_descriptor, ".?AVrange_error@std@@")
 DEFINE_CXX_DATA2(range_error, &runtime_error_cxx_type_info, &exception_cxx_type_info, MSVCP_runtime_error_dtor)
 
-#if _MSVCP_VER >= 100
+#if _MSVCP_VER >= 100 || _MSVCP_VER == 60
 /* overflow_error class data */
 typedef runtime_error overflow_error;
 extern const vtable_ptr overflow_error_vtable;
@@ -902,6 +902,109 @@ overflow_error* __thiscall overflow_error_copy_ctor(overflow_error *this, const
 DEFINE_RTTI_DATA2(overflow_error, 0, &runtime_error_rtti_base_descriptor, &exception_rtti_base_descriptor, ".?AVoverflow_error@std@@")
 DEFINE_CXX_DATA2(overflow_error, &runtime_error_cxx_type_info, &exception_cxx_type_info, MSVCP_runtime_error_dtor)
 
+#if _MSVCP_VER == 60
+/* ??0overflow_error@std@@QAE@ABV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@@Z */
+/* ??0overflow_error@std@@QEAA@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@@Z */
+DEFINE_THISCALL_WRAPPER(MSVCP_overflow_error_ctor_bstr, 8)
+overflow_error* __thiscall MSVCP_overflow_error_ctor_bstr(overflow_error *this, const basic_string_char *str)
+{
+    const char *name = MSVCP_basic_string_char_c_str(str);
+    TRACE("(%p %p %s)\n", this, str, name);
+    MSVCP_runtime_error_ctor(this, EXCEPTION_NAME(name));
+    this->e.vtable = &overflow_error_vtable;
+    return this;
+}
+
+/* ??4overflow_error@std@@QAEAAV01@ABV01@@Z */
+/* ??4overflow_error@std@@QEAAAEAV01@AEBV01@@Z */
+DEFINE_THISCALL_WRAPPER(MSVCP_overflow_error_assign, 8)
+overflow_error* __thiscall MSVCP_overflow_error_assign(overflow_error *this, const overflow_error *assign)
+{
+    MSVCP_runtime_error_dtor(this);
+    return overflow_error_copy_ctor(this, assign);
+}
+
+/* underflow_error class data */
+typedef runtime_error underflow_error;
+extern const vtable_ptr underflow_error_vtable;
+
+/* ??0underflow_error@std@@QAE@ABV01@@Z */
+/* ??0underflow_error@std@@QEAA@AEBV01@@Z */
+DEFINE_THISCALL_WRAPPER(underflow_error_copy_ctor, 8)
+underflow_error* __thiscall underflow_error_copy_ctor(underflow_error *this, const underflow_error *rhs)
+{
+    TRACE("%p %p\n", this, rhs);
+    runtime_error_copy_ctor(this, rhs);
+    this->e.vtable = &underflow_error_vtable;
+    return this;
+}
+
+/* ??0underflow_error@std@@QAE@ABV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@@Z */
+/* ??0underflow_error@std@@QEAA@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@@Z */
+DEFINE_THISCALL_WRAPPER(MSVCP_underflow_error_ctor_bstr, 8)
+underflow_error* __thiscall MSVCP_underflow_error_ctor_bstr(underflow_error *this, const basic_string_char *str)
+{
+    const char *name = MSVCP_basic_string_char_c_str(str);
+    TRACE("(%p %p %s)\n", this, str, name);
+    MSVCP_runtime_error_ctor(this, EXCEPTION_NAME(name));
+#ifndef SG_MUTANT_UNDERFLOW_AS_OVERFLOW
+    this->e.vtable = &underflow_error_vtable;
+#else
+    this->e.vtable = &overflow_error_vtable;
+#endif
+    return this;
+}
+
+/* ??4underflow_error@std@@QAEAAV01@ABV01@@Z */
+/* ??4underflow_error@std@@QEAAAEAV01@AEBV01@@Z */
+DEFINE_THISCALL_WRAPPER(MSVCP_underflow_error_assign, 8)
+underflow_error* __thiscall MSVCP_underflow_error_assign(underflow_error *this, const underflow_error *assign)
+{
+    MSVCP_runtime_error_dtor(this);
+    return underflow_error_copy_ctor(this, assign);
+}
+
+DEFINE_RTTI_DATA2(underflow_error, 0, &runtime_error_rtti_base_descriptor, &exception_rtti_base_descriptor, ".?AVunderflow_error@std@@")
+
+/* domain_error class data */
+typedef logic_error domain_error;
+extern const vtable_ptr domain_error_vtable;
+
+/* ??0domain_error@std@@QAE@ABV01@@Z */
+/* ??0domain_error@std@@QEAA@AEBV01@@Z */
+DEFINE_THISCALL_WRAPPER(domain_error_copy_ctor, 8)
+domain_error* __thiscall domain_error_copy_ctor(domain_error *this, const domain_error *rhs)
+{
+    TRACE("%p %p\n", this, rhs);
+    logic_error_copy_ctor(this, rhs);
+    this->e.vtable = &domain_error_vtable;
+    return this;
+}
+
+/* ??0domain_error@std@@QAE@ABV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@@Z */
+/* ??0domain_error@std@@QEAA@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@1@@Z */
+DEFINE_THISCALL_WRAPPER(MSVCP_domain_error_ctor_bstr, 8)
+domain_error* __thiscall MSVCP_domain_error_ctor_bstr(domain_error *this, const basic_string_char *str)
+{
+    const char *name = MSVCP_basic_string_char_c_str(str);
+    TRACE("(%p %p %s)\n", this, str, name);
+    MSVCP_logic_error_ctor(this, EXCEPTION_NAME(name));
+    this->e.vtable = &domain_error_vtable;
+    return this;
+}
+
+/* ??4domain_error@std@@QAEAAV01@ABV01@@Z */
+/* ??4domain_error@std@@QEAAAEAV01@AEBV01@@Z */
+DEFINE_THISCALL_WRAPPER(MSVCP_domain_error_assign, 8)
+domain_error* __thiscall MSVCP_domain_error_assign(domain_error *this, const domain_error *assign)
+{
+    MSVCP_logic_error_dtor(this);
+    return domain_error_copy_ctor(this, assign);
+}
+
+DEFINE_RTTI_DATA2(domain_error, 0, &logic_error_rtti_base_descriptor, &exception_rtti_base_descriptor, ".?AVdomain_error@std@@")
+#endif
+
 /* ?_Xoverflow_error@std@@YAXPBD@Z */
 /* ?_Xoverflow_error@std@@YAXPEBD@Z */
 void __cdecl DECLSPEC_NORETURN _Xoverflow_error(const char *str)
@@ -1706,11 +1809,19 @@ __ASM_BLOCK_BEGIN(exception_vtables)
     EXCEPTION_VTABLE(range_error,
             VTABLE_ADD_FUNC(MSVCP_runtime_error_vector_dtor)
             VTABLE_ADD_FUNC(MSVCP_runtime_error_what));
-#if _MSVCP_VER >= 100
+#if _MSVCP_VER >= 100 || _MSVCP_VER == 60
     EXCEPTION_VTABLE(overflow_error,
             VTABLE_ADD_FUNC(MSVCP_runtime_error_vector_dtor)
             VTABLE_ADD_FUNC(MSVCP_runtime_error_what));
 #endif
+#if _MSVCP_VER == 60
+    EXCEPTION_VTABLE(underflow_error,
+            VTABLE_ADD_FUNC(MSVCP_runtime_error_vector_dtor)
+            VTABLE_ADD_FUNC(MSVCP_runtime_error_what));
+    EXCEPTION_VTABLE(domain_error,
+            VTABLE_ADD_FUNC(MSVCP_logic_error_vector_dtor)
+            VTABLE_ADD_FUNC(MSVCP_logic_error_what));
+#endif
 
 __ASM_BLOCK_END
 
@@ -1771,9 +1882,13 @@ void init_exception(void *base)
     init_failure_rtti(base);
     init_bad_cast_rtti(base);
     init_range_error_rtti(base);
-#if _MSVCP_VER >= 100
+#if _MSVCP_VER >= 100 || _MSVCP_VER == 60
     init_overflow_error_rtti(base);
 #endif
+#if _MSVCP_VER == 60
+    init_underflow_error_rtti(base);
+    init_domain_error_rtti(base);
+#endif
 
     init_exception_cxx(base);
     init_bad_alloc_cxx(base);
@@ -1801,7 +1916,7 @@ void init_exception(void *base)
 #endif
     init_failure_cxx(base);
     init_range_error_cxx(base);
-#if _MSVCP_VER >= 100
+#if _MSVCP_VER >= 100 || _MSVCP_VER == 60
     init_overflow_error_cxx(base);
 #endif
 #endif
